text/0048-schema-annotations.md

```diff
@@ -0,0 +1,242 @@
+predict all of the metadata uses that users will have.
+Thus providing a flexible system that users can adapt to their needs is preferable.
+This proposal re-uses the same syntax from Cedar Policies, creating a unified syntax.
+
+
+## Detailed design
+
+### Semantics
+Attributes have **no** impact on validation decisions.
+Attributes are arbitrary key/value pairs where:
+* 'key' is a valid Cedar identifier
+* 'value' is a Cedar string
+
+
+The Cedar spec takes no opinion or stance on the interpretation of annotations.
+The interpretation is entirely up to users of Cedar.
+
+### Cedar Schema Syntax
+Attributes in Cedar Schemas will mirror the syntax used for attributes in a policy: informally that's `@<key>("value")`.
+Formally the following rule is added to the Cedar grammar:
+```
+Annotation := '@' IDENT '(' STR ')'
+Annotations := {Annotations}
+```
+With an arbitrary number of them being able to prepend to a namespace declaration, entity type declaration, common type declaration, action declaration, and an attribute declaration.
+
+Thus the full schema syntax becomes:
+```
+Schema      := {Namespace}
+Namespace   := (Annotations 'namespace' Path '{' {Decl} '}') | {Decl}
+Decl        := Entity | Action | TypeDecl
+Entity      := Annotations 'entity' Idents ['in' EntOrTyps] [['='] RecType] ';'
+Action      := Annotations 'action' Names ['in' (Name | '[' [Names] ']')] [AppliesTo] [ActAttrs]';'
+TypeDecl    := Annotations 'type' IDENT '=' Type ';'
+Type        := PRIMTYPE | IDENT | SetType | RecType
+EntType     := Path
+SetType     := 'Set' '<' Type '>'
+RecType     := '{' [AttrDecls] '}'
+AttrDecls   := Annotations Name ['?'] ':' Type [',' | ',' AttrDecls]
+AppliesTo   := 'appliesTo' '{' AppDecls '}'
+ActAttrs    := 'attributes' '{' AttrDecls '}'
+AppDecls    := ('principal' | 'resource') ':' EntOrTyps [',' | ',' AppDecls]
+             | 'context' ':' RecType [',' | ',' AppDecls]
+Path        := IDENT {'::' IDENT}
+EntTypes    := Path {',' Path}
+EntOrTyps   := EntType | '[' [EntTypes] ']'
+Name        := IDENT | STR
+Names       := Name {',' Name}
+Idents      := IDENT {',' IDENT}
+Annotation  := '@' IDENT '(' STR ')
+Annotations := Annotation {Annotations}
+
+IDENT       := ['_''a'-'z''A'-'Z']['_''a'-'z''A'-'Z''0'-'9']* - PRIMTYPE
+STR         := Fully-escaped Unicode surrounded by '"'s
+PRIMTYPE    := 'Long' | 'String' | 'Bool'
+WHITESPC    := Unicode whitespace
+COMMENT     := '//' ~NEWLINE* NEWLINE
+```
+
+### JSON Syntax
+None of the three top-level constructs (EntityTypes, Actions, CommonTypes) in JSON schemas allow for arbitrary key/value pairs.
+This means a new key can be safely added while preserving backwards compatibility.
+The same fact also applies to entity attribute declarations.
+This proposal reserves the `annotations` key at the top level of each of those constructs, which contains an Object, containing each annotation key as an Object key, associated with the annotation string.
```

Comment:
Maybe we could use `__annotations` if this is a concern

Reply:
Maybe we can also use reserved `__cedar::schema::annotations` to be 100% secure.